Ports/build_all.sh

```diff
@@ -1,89 +1,109 @@
 #!/usr/bin/env bash
 
-clean=false
-verbose=false
+some_failed='false'
+action='build'
+verbose='false'
+failfast='false'
 
-case "$1" in
-    clean)
-        clean=true
-        ;;
-    verbose)
-        verbose=true
-        ;;
-    *)
-        ;;
-esac
-
-case "$2" in
-    clean)
-        clean=true
-        ;;
-    verbose)
-        verbose=true
-        ;;
-    *)
-        ;;
-esac
+for arg in "$@"; do
+    case "$arg" in
+        clean*)
+            action="$arg"
+            ;;
+        verbose)
+            verbose=true
+            ;;
+        failfast)
+            failfast=true
+            ;;
+    esac
+done
 
 some_failed=false
-built_ports=""
+processed_ports=()
 
-for file in *; do
-    if [ -d $file ]; then
-        pushd $file > /dev/null
-            port=$(basename $file)
-            port_built=0
-            for built_port in $built_ports; do
-                if [ "$built_port" = "$port" ]; then
-                    port_built=1
-                    break
-                fi
-            done
-            if [ $port_built -eq 1 ]; then
-                echo "Built $port."
-                popd > /dev/null
-                continue
-            fi
-            if ! [ -f package.sh ]; then
-                echo "ERROR: Skipping $port because its package.sh script is missing."
-                popd > /dev/null
-                continue
-            fi
+log_success() {
+    echo -e "\033[1;32m[#]\033[0m $1"
+}
+log_warn() {
+    echo -e "\033[1;33m[!]\033[0m $1"
+}
+log_info() {
+    echo -e "\033[1;36m[*]\033[0m $1"
+}
+log_process() {
+    echo -e "\033[1m[~]\033[0m $1"
+}
+log_error() {
+    echo -e "\033[1;31m[x]\033[0m $1"
+}
+
+do_build_port() {
+    log_process "Building $port_name"
+    if $verbose; then
+        ./package.sh
+    else
+        ./package.sh &> /dev/null
+    fi
+}
 
-            if [ "$clean" == true ]; then
-                if [ "$verbose" == true ]; then
-                    ./package.sh clean_all
-                else
-                    ./package.sh clean_all > /dev/null 2>&1
+do_clean_port() {
+    log_process "Cleaning $port_name"
+    if $verbose; then
+        ./package.sh "$1"
+    else
+        ./package.sh "$1" &> /dev/null
+    fi
+}
+
+# It is required to simply implement cd, instead of pushd-popd
+ports_dir=$(realpath "$(dirname "${BASH_SOURCE[0]}")")
```

Comment:
> Using realpath of the ports directory to use `cd` operation only once and get rid of pushd-popd pattern here.

This will drop us into the directory of the last port rather than the directory where we started after the script is done, which I'd consider undesirable.

Reply:
I dont think so, because `cd` will happen only in context of script. You can tryout this snippet

```sh
ports_dir=$(realpath "$(dirname "${BASH_SOURCE[0]}")")
for port_dir in $(find "$ports_dir" -mindepth 1 -maxdepth 1 -type d | sort); do
    cd "$port_dir"
done
exit 1
```

![image](https://github.com/SerenityOS/serenity/assets/28386721/0c69ea84-2424-4159-9359-a263cf2f394f)

If you are talking about _in script it will drop us into the directory of last port_, then yes, it will. For this we can add a fix

```diff
+ old_pwd="$PWD"
ports_dir=$(realpath "$(dirname "${BASH_SOURCE[0]}")")
for port_dir in $(find "$ports_dir" -mindepth 1 -maxdepth 1 -type d | sort); do
    ...
done
if $some_failed; then
    exit 1
fi
+ cd "$old_pwd" || exit 1
```

Reply:
Ah, no, that is fine then. I've been spending too much time with `.ports_include.sh`, which needs to pay attention to that in most places.